File: arrays_and_hashing/0001_two_sum.py
# https://leetcode.com/problems/two-sum/description/
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        mapping = {}
        for idx, item in enumerate(nums):
            diff = target - item
            if diff in mapping:
                return [mapping[diff], idx]
            mapping[item] = idx

        # O(n^2), reasonably efficient on "small" nums array
        # for i in range(len(nums)):
        #     for j in range(i+1, len(nums)):
        #         if nums[i] + nums[j] == target:
        #             return [i, j]

File: arrays_and_hashing/0001_two_sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> mapping = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int diff = target - nums[i];

            if (mapping.containsKey(diff)) {
                return new int[] { mapping.get(diff), i };
            }

            mapping.put(nums[i], i);
        }

        return new int[0];
    }
}

// class Solution {
//     public int[] twoSum(int[] nums, int target) {
//         for (int i = 0; i < nums.length - 1; i++) {
//             for (int j = i + 1; j < nums.length; j++) {
//                 if (nums[i] + nums[j] == target) {
//                     return new int[]{i, j};
//                 }
//             }
//         }
//         return new int[]{};
//     }
// }
